File: Leetcode/Java Solutions/SlidingWindowMaximum.java
/*
LeetCode Problem: 239. Sliding Window Maximum
Link: https://leetcode.com/problems/sliding-window-maximum/
Written by: Mostofa Adib Shakib
Language: Java
Time Complexity: O(n)
Space Complexity: O(n)
*/

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> deque = new ArrayDeque<Integer>();
        int length = nums.length;
        int [] result = new int[length - k +1];
        int index = 0;
        
        for (int i = 0; i < length; i++) {
            if (!deque.isEmpty() && i - deque.peekFirst() == k) {
                deque.removeFirst();
            }
            
            while (!deque.isEmpty() && nums[deque.peekLast()] < nums[i]) {
                deque.removeLast();
            }
            
            deque.offer(i);
            
            if (i+1 >= k) {
                result[index] = nums[deque.peekFirst()];
                index++;
            }
        }
        
        return result;
    }
}

File: Leetcode/Python Solutions/Sliding Window/SlidingWindowMaximum.py
"""
LeetCode Problem: 239. Sliding Window Maximum
Link: https://leetcode.com/problems/sliding-window-maximum/
Language: Python
Written by: Mostofa Adib Shakib

Time Complexity: O(n)
Space Complexity: O(n)
"""

class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        
        deque = []      # Initializing deque
        result = []     # Initializing an array
        
        for i in range(len(nums)):
            # If elements at the top of the deque is outside the current window
            if deque and i - deque[0] == k:
                deque.pop(0)
            
            # If the last elements are smaller than the element being considered
            while deque and nums[deque[-1]] < nums[i]:
                deque.pop()
            
            deque.append(i)         # Push the index of the element to the deque
            
            # If we have one finished window
            if i+1 >= k:
                result.append(nums[deque[0]])
                
        return result